src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotOverloadOperatorEqualsOnReferenceTypes.cs

```diff
@@ -0,0 +1,60 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class DoNotOverloadOperatorEqualsOnReferenceTypes : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA1046";
+
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(MicrosoftCodeQualityAnalyzersResources.DoNotOverloadOperatorEqualsOnReferenceTypesTitle), MicrosoftCodeQualityAnalyzersResources.ResourceManager, typeof(MicrosoftCodeQualityAnalyzersResources));
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(MicrosoftCodeQualityAnalyzersResources.DoNotOverloadOperatorEqualsOnReferenceTypesMessage), MicrosoftCodeQualityAnalyzersResources.ResourceManager, typeof(MicrosoftCodeQualityAnalyzersResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftCodeQualityAnalyzersResources.DoNotOverloadOperatorEqualsOnReferenceTypesDescription), MicrosoftCodeQualityAnalyzersResources.ResourceManager, typeof(MicrosoftCodeQualityAnalyzersResources));
+
+        public static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Design,
+            RuleLevel.IdeSuggestion,
```

Comment:
@sharwell @stephentoub Any concerns in having this rule enabled as an IDE suggestion? Seems the rule will be correct majority of the times..

Reply:
There are plenty of places where someone overrides Equals on reference types.  Seems like this could have a lot of false positives.  We would definitely disable this in dotnet/runtime (e.g. Version.Equals), presumably in dotnet/roslyn (e.g. Symbol.Equals), presumably in dotnet/aspnetcore (e.g. BindingSource.Equals), etc.  Just doing a regex search for `class .* IEquatable<` to get a sense for how many reference types implement the interface (the number that override Equals is almost certainly more):
https://grep.app/search?q=class%20.%2A%20IEquatable%3C&regexp=true&case=true&filter[lang][0]=C%23

Reply:
@bartonjs, what's our current design guideline around this?

Reply:
* CONSIDER overriding Equals to provide value equality if a reference type represents a value. For example, you might want to override Equals in reference types representing numbers or other mathematical entities.
* DO NOT implement value equality on mutable reference types.

So it's right in the latter case, wrong in the former.  Maybe if it can reliably detect mutability?

Of course, that's for public types.  For private types you could want to match on a small number of fields and then have mutable hangers-on, like linked list pointers.

Reply:
@stephentoub Just in case, this is complaining about operator equal not method equals.

Reply:
> this is complaining about operator equal not method equals.

Ah, I misread the rule name, thanks.  Even so, providing an operator== is reasonable when you override Equals.

Reply:
Nevermind, that's for .Equals.  This is op_Equality.

* AVOID overloading equality operators on mutable reference types.
* AVOID overloading equality operators on reference types if the implementation would be significantly slower than that of reference equality.

I'm not exactly sure why we're softer on the operator than the method; apparently those sections are just far enough apart that it's been a cache-miss for 3 editions of the guidelines :)

Reply:
> AVOID overloading equality operators on mutable reference types.

I guess if any field isn't `readonly` or any property has a setter we could assume this is a mutable type.

> AVOID overloading equality operators on reference types if the implementation would be significantly slower than that of reference equality.

I don't really see how we could do so...

> There's a CONSIDER rule for overriding op_Equality for anything that implements IEquatable<TheirOwnType>.

Maybe the best is to say report on classes except if it implements `IEquatable`, how does it sound to you?

Reply:
@Evangelink, I think it would be best to start with this rule as disabled by default. We can move the discussion around whether it would be suitable to be enabled as an IDE suggestion by default to a separate issue.